src/Symfony/Component/Security/Core/Authorization/AccessDecisionManager.php

```diff
@@ -28,30 +32,34 @@ class AccessDecisionManager implements AccessDecisionManagerInterface
 
     private $voters;
     private $strategy;
-    private $allowIfAllAbstainDecisions;
-    private $allowIfEqualGrantedDeniedDecisions;
+    private $voteStrategy;
 
     /**
      * Constructor.
      *
-     * @param VoterInterface[] $voters                             An array of VoterInterface instances
-     * @param string           $strategy                           The vote strategy
-     * @param bool             $allowIfAllAbstainDecisions         Whether to grant access if all voters abstained or not
-     * @param bool             $allowIfEqualGrantedDeniedDecisions Whether to grant access if result are equals
+     * @param VoterInterface[]             $voters                             An array of VoterInterface instances
+     * @param string|VoteStrategyInterface $strategy                           The vote strategy
+     * @param bool                         $allowIfAllAbstainDecisions         Whether to grant access if all voters abstained or not
+     * @param bool                         $allowIfEqualGrantedDeniedDecisions Whether to grant access if result are equals
      *
      * @throws \InvalidArgumentException
      */
     public function __construct(array $voters = array(), $strategy = self::STRATEGY_AFFIRMATIVE, $allowIfAllAbstainDecisions = false, $allowIfEqualGrantedDeniedDecisions = true)
     {
-        $strategyMethod = 'decide'.ucfirst($strategy);
-        if (!is_callable(array($this, $strategyMethod))) {
-            throw new \InvalidArgumentException(sprintf('The strategy "%s" is not supported.', $strategy));
-        }
-
         $this->voters = $voters;
-        $this->strategy = $strategyMethod;
-        $this->allowIfAllAbstainDecisions = (bool) $allowIfAllAbstainDecisions;
-        $this->allowIfEqualGrantedDeniedDecisions = (bool) $allowIfEqualGrantedDeniedDecisions;
+        $this->strategy = $strategy;
+
+        if ($strategy instanceof VoteStrategyInterface) {
+            $this->voteStrategy = $strategy;
+        } elseif (self::STRATEGY_AFFIRMATIVE === $strategy) {
```

Comment:
And maybe removing `allowIfAllAbstainDecisions`, `allowIfEqualGrantedDeniedDecisions` from constructor of this class in future?